### PR TITLE
Adding a metadata file to KubeHound dumps

### PR DESCRIPTION
**Pull request by edznux-dd**
This is barely a PoC.

TLDR:
To simplify user experience, especially for the local dumper/ingestor, we should write a `metadata.json` file that contains the metadata of a specific dump.

Why:
Stuffing data in the filepath as some limitation:
- There's a limit of number of element in the data path (limited by the number of characters for each filename)
- Force the user to never rename the archive, which can be an issue when storing historical dumps (currently, the latest dump is computed by the file last modified time)
- Parsing the data can become more cumbersome than simply parsing a json file containing each value in the correct format.

How:
Add a `metadata.json` file that contains all the metadata around a specific dump.
It will then be reused by the ingestor to populate the RunID, the ClusterName and any future info we want.

### CODE REVIEW

File: pkg/collector/file.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ const (
 	PodPath                 = "pods.json"
 	RolesPath               = "roles.rbac.authorization.k8s.io.json"
 	RoleBindingsPath        = "rolebindings.rbac.authorization.k8s.io.json"
+	MetadatPath             = "metadata.json"
```
Comment (edznux-dd) [inline pkg/collector/file.go:48]: typo

File: pkg/kubehound/core/core_ingest_local.go [hunk 2]
```diff
@@ -22,7 +24,7 @@ func CoreLocalIngest(ctx context.Context, khCfg *config.KubehoundConfig, resultP
 	if err != nil {
 		return err
 	}
-
+	metadataFilePath := filepath.Join(resultPath, "metadata.json")
```
Comment (edznux-dd) [inline pkg/kubehound/core/core_ingest_local.go:27]: Should we reuse the global variable there as well?

File: pkg/dump/writer/file_writer.go [hunk 3]
```diff
@@ -56,6 +56,10 @@ func (f *FileWriter) WorkerNumber() int {
 	return FileWriterWorkerNumber
 }
 
+func (f *FileWriter) WriteMetadata(ctx context.Context) error {
+	return nil
+}
```
Comment (edznux-dd) [inline pkg/dump/writer/file_writer.go:61]: reomve this method and the next one since you moved it to the pipeline

File: pkg/kubehound/core/core_ingest_local.go [hunk 4]
```diff
@@ -35,6 +37,11 @@ func CoreLocalIngest(ctx context.Context, khCfg *config.KubehoundConfig, resultP
 		if err != nil {
 			return err
 		}
+		metadataFilePath = filepath.Join(tmpDir, "metadata.json")
```
Comment (edznux-dd) [inline pkg/kubehound/core/core_ingest_local.go:40]: same here

File: pkg/dump/pipeline/metadata.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+import (
+	"context"
+
+	"github.com/DataDog/KubeHound/pkg/collector"
+	"github.com/DataDog/KubeHound/pkg/dump/writer"
+)
+
+type MetadataIngestor struct {
+	buffer map[string]*collector.Metadata
```
Comment (edznux-dd) [inline pkg/dump/pipeline/metadata.go:11]: Do we need that to be a pointer of `collector.Metadata` instead of simply `map[string]collector.Metadata` ?

File: pkg/collector/k8s_api.go [hunk 6]
```diff
@@ -114,19 +116,35 @@ func NewK8sAPICollector(ctx context.Context, cfg *config.KubehoundConfig) (Colle
...
-		rl:        ratelimit.New(cfg.Collector.Live.RateLimitPerSecond), // per second
-		tags:      *newCollectorTags(),
-		waitTime:  map[string]time.Duration{},
-		startTime: time.Now(),
-		mu:        &sync.Mutex{},
+		cfg:         cfg.Collector.Live,
+		clientset:   clientset,
+		log:         l,
+		rl:          ratelimit.New(cfg.Collector.Live.RateLimitPerSecond), // per second
+		tags:        *newCollectorTags(),
```
Comment (edznux-dd) [inline pkg/collector/k8s_api.go:123]: same here, do we need a pointer to this instead of just returning the res? :shrug:

File: pkg/dump/pipeline/metadata.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+func NewMetadataIngestor(ctx context.Context, dumpWriter writer.DumperWriter) *MetadataIngestor {
+	return &MetadataIngestor{
+		buffer: make(map[string]collector.Metadata),
+		writer: dumpWriter,
+	}
+}
+
+func (d *MetadataIngestor) DumpMetadata(ctx context.Context, metadata collector.Metadata) error {
+	data := make(map[string]*collector.Metadata)
```
Comment (edznux-dd) [inline pkg/dump/pipeline/metadata.go:23]: Is it expected that this is a pointer to the collector metadata ? :sweat_smile:

The buffer above is `make(map[string]collector.Metadata),`

File: pkg/kubehound/core/core_ingest_local.go [hunk 8]
```diff
@@ -4,16 +4,20 @@ import (
...
+	"github.com/DataDog/KubeHound/pkg/dump"
 	"github.com/DataDog/KubeHound/pkg/ingestor/puller"
 )
 
 func CoreLocalIngest(ctx context.Context, khCfg *config.KubehoundConfig, resultPath string) error {
 	// Using the collector config to ingest the data
 	khCfg.Collector.Type = config.CollectorTypeFile
 	// Applying the clusterName from the dynamic config (from CLI or env var) to the collector config
-	khCfg.Collector.File.ClusterName = khCfg.Dynamic.ClusterName
+	// khCfg.Collector.File.ClusterName = khCfg.Dynamic.ClusterName
```
Comment (edznux-dd) [inline pkg/kubehound/core/core_ingest_local.go:19]: is this wanted? (should it remove the comment altogether instead?)

File: pkg/dump/result.go [hunk 9]
```diff
@@ -70,44 +74,34 @@ func (i *DumpResult) Compressed() {
...
 }
 
-func ParsePath(path string) (*DumpResult, error) {
-	// ./<clusterName>/kubehound_<clusterName>_<run_id>[.tar.gz]
-	// re := regexp.MustCompile(`([a-z0-9\.\-_]+)/kubehound_([a-z0-9\.-_]+)_([a-z0-9]{26})\.?([a-z0-9\.]+)?`)
-	re := regexp.MustCompile(DumpResultPathRegex)
-	if !re.MatchString(path) {
-		return nil, fmt.Errorf("Invalid path provided: %q", path)
-	}
+func GetDumpMetadata(ctx context.Context, metadataFilePath string) (collector.Metadata, error) {
```
[the remark this answers is not in the document]
Comment (jt-dd) [reply pkg/dump/result.go:89]: as you want, I used yours 🙃 , will change to `ParseMetadata`

File: pkg/kubehound/core/core_ingest_local.go (hunk 8 above)
Comment (jt-dd) [reply pkg/kubehound/core/core_ingest_local.go:19]: yes will remove the comment also

File: pkg/dump/pipeline/metadata.go (hunk 7 above)
Comment (jt-dd) [reply pkg/dump/pipeline/metadata.go:23]: no it is not 😅

File: pkg/kubehound/core/core_ingest_local.go (hunk 4 above)
Comment (jt-dd) [reply pkg/kubehound/core/core_ingest_local.go]: done

File: pkg/dump/pipeline/metadata.go (hunk 5 above)
Comment (jt-dd) [reply pkg/dump/pipeline/metadata.go]: done

File: pkg/collector/k8s_api.go (hunk 6 above)
Comment (jt-dd) [reply pkg/collector/k8s_api.go:123]: fixed

**Review by jt-dd — APPROVED** [review]
